### DISCUSSION TITLE
route_callback

### DISCUSSION THREAD

**Discussion by tlzfighting** [contiki-ng/contiki-ng#2328] [General]
### In the hello-world.c , I include"   #include "os/net/ipv6/uip-ds6-route.h"   " to   implement routing callback function

static struct uip_ds6_notification n;

![图片](https://user-images.githubusercontent.com/95647449/206714401-fe53788c-a580-4494-a69b-b470b3d933a1.png)
 
uip_ds6_notification_add(&n, route_callback);

### **But  when I use command  " make TARGET=cc2538dk " ,The following issues emerged.  How can I solve this?Looking forward to your valuable reply!**

hello-world.c: In function 'route_callback':
hello-world.c:57:15: error: 'UIP_DS6_NOTIFICATION_DEFRT_ADD' undeclared (first use in this function)
   if(event == UIP_DS6_NOTIFICATION_DEFRT_ADD) {
               ^
hello-world.c:57:15: note: each undeclared identifier is reported only once for each function it appears in
hello-world.c: In function 'process_thread_hello_world_process':
hello-world.c:102:3: error: implicit declaration of function 'uip_ds6_notification_add' [-Werror=implicit-function-declaration]
   uip_ds6_notification_add(&n, route_callback);
   ^
cc1.exe: all warnings being treated as errors
make: *** [../../Makefile.include:498：build/cc2538dk/obj/hello-world.o] error  1

**Comment by arurke** [comment]
Hi, @tlzfighting, I am converting this to a discussion. Please use gitter or [discussions](https://github.com/contiki-ng/contiki-ng/discussions) for these type of questions, ref. [the doc.](https://docs.contiki-ng.org/en/develop/doc/project/Where-to-report-issues,-ask-questions,-etc.html?highlight=issue#github-issues).

**Comment by arurke** [comment]
See `uip-ds6-route.h`: `UIP_DS6_NOTIFICATION_DEFRT_ADD` is only defined if `UIP_DS6_NOTIFICATIONS` is 1 - and this is only the case if you are running RPL classic (or if you manually set it). RPL lite is the default. To learn more about this and how to switch to RPL classic, check out [this doc. page](https://docs.contiki-ng.org/en/develop/doc/programming/RPL.html?highlight=rpl%20classic#rpl-classic)

**Reply by tlzfighting** [reply]
Hi，sir !  
     The function I want to achieve is to collect temperature data from the end node and then send it to the server via udp, so this routing callback function can be deleted, right, because it does not affect the function I want to achieve.
### The approximate process is as follows:

  static struct uip_ds6_notification n;
  
  static void
	route_callback(int event, uip_ipaddr_t *route, uip_ipaddr_t *ipaddr,
				   int numroutes)
	{
	  if(event == UIP_DS6_NOTIFICATION_DEFRT_ADD) {
	   // leds_off(LEDS_ALL);
		printf("Got a RPL route\n");
	  }
	}
	  
  uip_ds6_notification_add(&n, route_callback);
  

  udp_socket_register(&s, NULL, receiver);

  /* Bind UDP socket to local port */
 
  udp_socket_bind(&s, PORT);

  
  udp_socket_connect(&s, NULL, PORT);
  
   udp_socket_sendto(&s,
                      temp, 7,
                      &ip6addr, PORT);

### So can I delete the following content:
  static struct uip_ds6_notification n;
  
  static void
	route_callback(int event, uip_ipaddr_t *route, uip_ipaddr_t *ipaddr,
				   int numroutes)
	{
	  if(event == UIP_DS6_NOTIFICATION_DEFRT_ADD) {
	   // leds_off(LEDS_ALL);
		printf("Got a RPL route\n");
	  }
	}
	  
  uip_ds6_notification_add(&n, route_callback);

**Looking forward to your valuable reply!**
